### PR TITLE
feat(server): async request lifecycle, strict request schema, and execution guarantees

### PR DESCRIPTION
**Pull request by Yunnglin**
## Summary

Hardens the Twinkle server's entire request-serving path across three sequenced specs, then enforces lint on the client and lands the supporting refactors. Client and server share one request-model contract; every serving path has a strict time bound; failures propagate loudly.

- **Execution guarantees** — every serving path has a strict time upper bound; failures propagate as a complete `ErrorPayload` instead of degrading silently; a shared pydantic contract base.
- **Async submit/retrieve lifecycle (Part 2)** — twinkle-native requests submit work and return a `TaskEnvelope`; results are delivered through a persistent future record + retrieve endpoint, decoupled from the HTTP connection.
- **Request schema (Part 3)** — one shared request-model source of truth, a strict control plane, and a data-plane Wire Schema validated at HTTP ingress before anything reaches a GPU.

### 1. Execution guarantees (time bounds + loud failures)
- Effective time bound on the `sync=True` dispatch path; blocking calls moved off the event loop via `call_backend` + a per-replica admission gate; AST guard forbids new direct backend calls.
- Post-timeout actor liveness probe + health bit; two distinct bounds (record-terminal vs. resource-release) documented.
- Removed all silent-degradation layers and the `TWINKLE_FAIL_FAST` switch; `nccl_safe` reduced to a single `nccl_safe_megatron` decorator.
- Sampler streams gain a double timeout (total + per-token); failures surface as `ErrorPayload` wire frames.

### 2. Async submit/retrieve lifecycle (Part 2)
- `run_submit` / `submit_and_peek` submit shell: handlers submit and return immediately; the client future layer resolves the envelope to a terminal state.
- Persistent `FutureRecord` (with `replica_id` / `absolute_deadline`); cleanup reads each record's own deadline; do-not-regress guard extended to terminal->terminal.
- Idempotent `seq_id` dedup keyed on `(session, adapter, seq_id)` so a retried gradient-mutating call applies at most once.

### 3. Request schema (Part 3)
- Single shared request models in `twinkle_client.types`; the client builds them via `build_request` + one `model_dump_json(exclude_none=True)` instead of hand-assembling `json_data`.
- `StrictRequest` on all twinkle-native routes (unknown top-level field -> 422); `ResponseModel` (`extra='ignore'`) so older clients tolerate new response fields; `token` is rejected in the body.
- Explicit field roles (control / `backend_kwarg` / passthrough): each backend field forwarded at most once, control fields never forwarded, passthrough regions flattened and forwarded unjudged (no spelling heuristic).
- `run_submit` preflight rejects backend-incompatible fields (422) and unavailable endpoints (501) before the seq claim / enqueue -> a rejected request runs on **zero** data-parallel ranks.
- Wire Schema validates inline `inputs` at HTTP ingress: strict integer leaves (reject bool/float), homogeneous batch, unknown-JSON-field preservation on export, shallowest-first unions; a single shared `is_encoded` predicate replaces three copies.
- Unified `RequestValidationError -> ErrorPayload` handler registered on the shared deployment app builder (Model / Sampler / Processor identical).
- Naming disambiguation: sampler-domain models are `Sampler`-prefixed so they no longer collide with `model.py`; fixes `add_adapter_to_sampler` silently binding the wrong (model) schema and rejecting a dict `config`.

### 4. Lint enforcement & refactors
- Remove `src/twinkle_client/` from the pre-commit exemptions; the client is now linted like the server, and the existing client code is normalized to green.
- Rename `server/utils/validation.py -> auth.py` (ends the collision with the `server/validation` preflight package) and `model/utils.py -> data_plane_inputs.py`.
- Rename client `http/http_utils.py -> http/client.py` and `http/utils.py -> http/context.py`; update all importers.
- Split `auto/agent/tools.py` into `tool_schemas.py` / `server_tools.py` / `search_tools.py`; rework `rollout/multi_turn.py` into an explicit `_RolloutState`.

## Breaking changes
- Control-plane strictness and the Wire Schema are **breaking wire changes**: unknown top-level fields and previously-tolerated bool/float/heterogeneous inline inputs are now rejected. Client and server must be upgraded in **lockstep** — no version negotiation or lax mode is introduced.

## Testing
- **Schema:** preflight (422/501 with no future record and no backend call), request-wire (StrictRequest, `token` rejection, `seq_id` acceptance, sampler-binding regression), wire-schema (classification, homogeneity, strict ints, extension round-trip, `CORE_INPUT_KEYS`/`VLM_CONCAT_FIELDS` consistency).
- **Execution guarantees / lifecycle:** error-wire, stream guarantees, `nccl_safe`, blocking-boundary + actor-recovery integration, future-lifecycle / error-payload state tests, and the zero-wire-change contract baseline guard.
- `pre-commit run --all-files` green across server + client; local unit suites green in the vLLM env.

### CODE REVIEW

File: src/twinkle/protocol/types/model.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+
+class ForwardOnlyRequest(_InlineForwardBase):
+    """Body of ``POST /twinkle/forward_only``: no graph, no gradients.
+
+    ``adapter_name`` is optional here -- a reference forward may run against the base
+    weights -- and there is no ``seq_id`` because nothing is mutated to be idempotent
+    about.
+    """
+
+    adapter_name: str | None = None
```
[the remark this answers is not in the document]
Comment (Yunnglin) [reply src/twinkle/protocol/types/model.py]: Fixed in b4776b7. The backend requires an adapter for template and adapter context, including base-weight inference via disable_lora, so ForwardOnlyRequest now requires adapter_name.

File: src/twinkle/protocol/types/model.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+
+
+class AddAdapterRequest(StrictRequest):
+    adapter_name: str
+    # ``config`` is None for full-parameter training (no LoRA adapter) and a
+    # serialized LoraConfig string for LoRA training.
+    config: str | None = None
+    save_dir: str | None = None
+    gradient_accumulation_steps: int | None = backend_kwarg(default=None, ge=1)
+    init_kwargs: dict[str, JsonValue] = passthrough()
```
[the remark this answers is not in the document]
Comment (Yunnglin) [reply src/twinkle/protocol/types/model.py:314]: Fixed in b4776b7. The public client now accepts Mapping input, converts it to LoraConfig, and serializes it into the strict string wire field; regression coverage was added.

File: src/twinkle_client/auto/agent/core.py [hunk 3]
```diff
@@ -28,7 +28,7 @@ class AgentLoop:
     def __init__(
         self,
         connection: LocalConnection,
-        llm_client: 'AsyncOpenAI',
+        llm_client: AsyncOpenAI,
```
[the remark this answers is not in the document]
Comment (Yunnglin) [reply src/twinkle_client/auto/agent/core.py:34]: Fixed in b4776b7 by adding the TYPE_CHECKING-only AsyncOpenAI import.

File: src/twinkle_client/auto/agent/monitor.py [hunk 4]
```diff
@@ -95,7 +95,7 @@ def __init__(
         self,
         connection: LocalConnection,
         on_message: Callable[[str], None],
-        llm_client: 'AsyncOpenAI',
+        llm_client: AsyncOpenAI,
```
[the remark this answers is not in the document]
Comment (Yunnglin) [reply src/twinkle_client/auto/agent/monitor.py:101]: Fixed in b4776b7 by adding the TYPE_CHECKING-only AsyncOpenAI import.

File: src/twinkle_client/http/client.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+    def url(self, path_or_url: str = '') -> str:
+        if path_or_url.startswith(('http://', 'https://')):
+            return path_or_url
+        if not path_or_url:
+            return self._context.base_url
+        return f'{self._context.base_url}/{path_or_url.lstrip("/")}'
+
+    def _headers(self, additional_headers: Mapping[str, str] | None = None) -> dict[str, str]:
+        headers = build_routing_headers(self._context.routing_id, f'Bearer {self._context.api_key}')
```
[the remark this answers is not in the document]
Comment (Yunnglin) [reply src/twinkle_client/http/client.py:158]: Not reproducible. ClientTransport._headers uses the configured API key as Bearer auth; test_context_normalizes_url_and_transport_builds_stable_headers covers it.

File: src/twinkle/server/config/persistence.py [hunk 6]
```diff
@@ -22,16 +21,13 @@ class PersistenceConfig(BaseModel):
 
     model_config = ConfigDict(extra='forbid')
 
-    mode: Literal['memory', 'file', 'redis'] = 'memory'
-    file_path: str | None = None  # required for file mode
+    mode: Literal['memory', 'redis'] = 'memory'
```
[the remark this answers is not in the document]
Comment (Yunnglin) [reply src/twinkle/server/config/persistence.py:24]: Intentional breaking configuration change: file persistence is removed and only memory/redis remain. The current contract-hardening scope does not require migration documentation.

File: tests/data_format/test_tq_fields.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+required by (empty rows, inconsistent fields, numeric+non-numeric mix), plus a
+subprocess assertion (not depending on the ``async-rl`` extra) that guards: the
+module must import cleanly without ``tensordict``.
+"""
+import subprocess
+import sys
+
+import pytest
+
+tensordict = pytest.importorskip('tensordict')
```
[the remark this answers is not in the document]
Comment (Yunnglin) [reply tests/data_format/test_tq_fields.py]: Fixed in b4776b7. importorskip now runs only in the TensorDict-dependent tests, so the dependency-light import test executes without tensordict.

File: src/twinkle/data_format/encoding.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+from __future__ import annotations
+
+from collections.abc import Mapping
+from typing import Any
+
+# Presence of any of these keys means the entry carries encoded model input.
+ENCODED_INPUT_KEYS: tuple[str, ...] = ('input_ids', 'input_embedding')
+
+
+def is_encoded(entry: Any) -> bool:
```
Comment (tastelikefeet) [inline src/twinkle/data_format/encoding.py:24]: 这个看起来不是 data_format，迁移 utils？

File: src/twinkle/infra/_ray/ray_helper.py [hunk 9]
```diff
@@ -137,10 +137,17 @@ def is_worker():
...
+
+        ``timeout`` is passed to ``ray.get(list, timeout=)``, whose semantics are
+        the **total** wall-clock time to collect the whole list -- different from
+        ``LazyCollect``'s per-future timing (see ``do_get_and_collect_func``).
+        The two paths are each bounded on their own; the total-time semantics here
+        are strictly tighter.
+        """
         import ray
-        return ray.get(RayHelper.execute_all_async(method_name, workers_and_args))
+        return ray.get(RayHelper.execute_all_async(method_name, workers_and_args), timeout=timeout)
```
Comment (tastelikefeet) [inline src/twinkle/infra/_ray/ray_helper.py:150]: 如果增加 timeout，内部 hang 住会不会导致后续请求卡死

File: src/twinkle/model/megatron/megatron.py [hunk 10]
```diff
@@ -420,62 +419,47 @@ def forward_step_func(data_iterator, model):
...
                     if entropies is not None:
-                        _outputs['entropies'] = entropies
-                    if hasattr(_loss_instance, 'require_logits') and _loss_instance.require_logits:
-                        _outputs['logits'] = output_tensor
-                    batch, _outputs = processor.unpack_packed_sequences(batch, _outputs)
-                    logps = _outputs['logps']
-                    entropies = _outputs.get('entropies', None)
-                    unpacked_logits = _outputs.get('logits', None)
-            except Exception as e:
-                # Data processing error (e.g. unpack_packed_sequences dimension mismatch).
```
Comment (tastelikefeet) [inline src/twinkle/model/megatron/megatron.py:467]: 这里去掉 try-except 的原因呢

File: src/twinkle/model/optimizer_group.py [hunk 11]
```diff
@@ -48,12 +48,6 @@ class BaseOptimizerGroup:
     _device_mesh: DeviceMesh = None
     _last_grad_norm: float = 0.0
 
-    def __setattr__(self, name, value):
-        if name == 'loss_instance' and value is not None:
```
Comment (tastelikefeet) [inline src/twinkle/model/optimizer_group.py:52]: 这个去掉的原因呢

File: src/twinkle/model/megatron/megatron.py (hunk 10 above)
Comment (Yunnglin) [reply src/twinkle/model/megatron/megatron.py:467]: 以前出错会偷偷吞掉，再返回零 loss 假装训练成功；现在改成直接报错，让调用方知道训练确实失败了

File: src/twinkle/model/optimizer_group.py (hunk 11 above)
Comment (Yunnglin) [reply src/twinkle/model/optimizer_group.py:52]: 同上，同样是删掉“出错就用零 loss 顶替”的逻辑

File: src/twinkle/data_format/encoding.py (hunk 8 above)
Comment (Yunnglin) [reply src/twinkle/data_format/encoding.py:24]: 这是判断输入是不是“模型已经编码好的数据”，属于数据格式逻辑，不是通用工具；放 data_format 可能更合适

File: src/twinkle/infra/_ray/ray_helper.py (hunk 9 above)
Comment (Yunnglin) [reply src/twinkle/infra/_ray/ray_helper.py:150]: 这里机制链路是：

1. 服务端把调用发给多个 Ray actor，得到一组 future。  
2. `ray.get(futures, timeout=T)` 最多等 `T` 秒。  
3. 所有 actor 及时返回，就正常汇总结果；任意 actor 卡住导致整批未完成，就抛出超时异常。  
4. 队列捕获异常，把该请求写成失败终态，客户端轮询 future 时拿到超时错误。  
5. 超时后系统会检查模型还活不活；如果上一条任务其实还卡着，就先拒绝新请求，直接告诉用户“后端忙”，避免大家一起排队卡住。  
6. 等卡住的任务自己结束后，系统会自动恢复，新请求可以继续进来。

核心是：不能保证立刻杀死卡住的 actor，但能保证调用方不会无限等待，也不会让后续请求无止境堆积。默认执行 timeout 是 1800 秒（30 分钟）
